modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScript.java

```diff
@@ -0,0 +1,144 @@
+        // create a script stack: this is just the script portion
+        List<String> scriptStack = new ArrayList<>();
+        for (StackTraceElement element : t.getStackTrace()) {
+            if (WriterConstants.CLASS_NAME.equals(element.getClassName())) {
+                // found the script portion
+                int offset = element.getLineNumber();
+                if (offset == -1) {
+                    scriptStack.add("<<< unknown portion of script >>>");
+                } else {
+                    offset--; // offset is 1 based, line numbers must be!
+                    int startOffset = metadata.getPreviousStatement(offset);
+                    if (startOffset == -1) {
+                        assert false; // should never happen unless we hit exc in ctor prologue...
+                        startOffset = 0;
+                    }
+                    int endOffset = metadata.getNextStatement(startOffset);
+                    if (endOffset == -1) {
+                        endOffset = metadata.getSource().length();
+                    }
+                    // TODO: if this is still too long, truncate and use ellipses
+                    String snippet = metadata.getSource().substring(startOffset, endOffset);
+                    scriptStack.add(snippet);
+                    StringBuilder pointer = new StringBuilder();
+                    for (int i = startOffset; i < offset; i++) {
+                        pointer.append(' ');
+                    }
+                    pointer.append("^---- HERE");
+                    scriptStack.add(pointer.toString());
+                }
+                break;
+            // but filter our own internal stacks (e.g. indy bootstrap)
+            } else if (!shouldFilter(element)) {
+                scriptStack.add(element.toString());
+            }
+        }
+        // build a name for the script:
+        final String name;
+        if (PainlessScriptEngineService.INLINE_NAME.equals(metadata.getName())) {
+            name = metadata.getSource();
+        } else {
+            name = metadata.getName();
+        }
+        ScriptException scriptException = new ScriptException("runtime error", t, scriptStack, name, PainlessScriptEngineService.NAME);
+        for (Map.Entry<String, List<String>> entry : extraMetadata.entrySet()) {
+            scriptException.addMetadata(entry.getKey(), entry.getValue());
+        }
+        return scriptException;
+    }
+
+    /** returns true for methods that are part of the runtime */
+    private static boolean shouldFilter(StackTraceElement element) {
+        return element.getClassName().startsWith("org.elasticsearch.painless.") ||
+               element.getClassName().startsWith("java.lang.invoke.") ||
+               element.getClassName().startsWith("sun.invoke.");
+    }
+
+    /**
+     * Metadata about the script.
+     */
+    public static class ScriptMetadata {
```

Comment:
I'll move it back. When I had this implementing arbitrarily named methods it made sense to try and minimize the number of methods on the superclass.